### PR TITLE
Record ctime for underreplicated ledger mark time.

### PR DESCRIPTION
**Pull request by reddycharan**
Descriptions of the changes in this PR:

Enable the Auditor to use system time as underreplicated ledger mark time.
If this is enabled, Auditor will write a ctime field into the
underreplicated ledger znode. This would be useful in knowing when the
ledger is marked underreplicated, using BookieShell commands and also later
we can add alerts/metrics on under replicated ledgers which remained under
replicated for more than acceptable duration.

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/ZkLedgerUnderreplicationManager.java [hunk 1]
```diff
@@ -393,11 +390,17 @@ public boolean hasNext() {
...
-                                    List<String> replicaList = getLedgerUnreplicationInfo(ledgerId).getReplicaList();
-                                    if ((predicate == null)
-                                            || predicate.test(replicaList)) {
-                                        curBatch.add(new AbstractMap.SimpleImmutableEntry<Long, List<String>>(ledgerId,
-                                                ((includeReplicaList) ? replicaList : null)));
+                                    UnderreplicatedLedgerFormat underreplicatedLedgerFormat =
+                                            getLedgerUnreplicationInfo(ledgerId);
+                                    List<String> replicaList = underreplicatedLedgerFormat.getReplicaList();
+                                    long ctime = (underreplicatedLedgerFormat.hasCtime()
+                                            ? underreplicatedLedgerFormat.getCtime() : -1L);
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/ZkLedgerUnderreplicationManager.java]: What about having a constant for -1 (unknown) ?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java [hunk 2]
```diff
@@ -931,14 +932,17 @@ int runCmd(CommandLine cmdLine) throws Exception {
...
-                        .listLedgersToRereplicate(predicate, printMissingReplica);
+                Iterator<UnderreplicatedLedger> iter = underreplicationManager.listLedgersToRereplicate(predicate);
                 while (iter.hasNext()) {
-                    Map.Entry<Long, List<String>> urLedgerMapEntry = iter.next();
-                    long urLedgerId = urLedgerMapEntry.getKey();
+                    UnderreplicatedLedger underreplicatedLedger = iter.next();
+                    long urLedgerId = underreplicatedLedger.getLedgerId();
                     System.out.println(ledgerIdFormatter.formatLedgerId(urLedgerId));
+                    long ctime = underreplicatedLedger.getCtime();
+                    if (ctime != -1) {
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java]: Better to have a constant for -1

**Comment by reddycharan** [comment]
@jvrao @athanatos can you take a look

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java (hunk 2 above)
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java]: +1 to have a constant

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerUnderreplicationManager.java [hunk 3]
```diff
@@ -29,6 +28,40 @@
  * Interface for marking ledgers which need to be rereplicated.
  */
 public interface LedgerUnderreplicationManager extends AutoCloseable {
+
+    /**
+     * UnderReplicated ledger representation info.
+     */
+    class UnderreplicatedLedger {
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerUnderreplicationManager.java]: move this out as a separate class file in this package

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java (hunk 2 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java]: done

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/ZkLedgerUnderreplicationManager.java (hunk 1 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/ZkLedgerUnderreplicationManager.java]: done

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerUnderreplicationManager.java (hunk 3 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerUnderreplicationManager.java]: done

**Comment by reddycharan** [comment]
retest this please

**Pull request closed without merging**